File: UnionFind/1976.py
# 11:27
# 
# 1 -> 1-2
# 2 -> 1-2 & 1-3
# 3 -> 2-3 

def find(x):
    if parent[x] != x:
        return find(parent[x])
    return x

def union(a,b):
    a = find(a)
    b = find(b)
    
    if a < b:
        parent[b] = a
    else:
        parent[a] = b

n = int(input())
m = int(input())

parent = []
for a in range(n+1):
    parent.append(a)

orders = []
for _ in range(n):
    orders.append(list(map(int, input().split())))

l = list(map(int, input().split()))

for a in range(len(orders)):
    for b in range(len(orders[a])):
        if orders[a][b] == 1:
            union(a+1, b+1)

current = -1
check = True
for a in range(len(l)):
    if a == 0:
        if parent[l[a]] != l[a]:
            parent[l[a]] = find(l[a])
        current = parent[l[a]]
    else:
        if parent[l[a]] != l[a]:
            parent[l[a]] = find(l[a])
        
        if parent[l[a]] != current:
            check = False
            break

if check:
    print("YES")
else:
    print("NO")

File: C++/1976.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;

vector<int> parent;
int n,m;
int a,b,c;
vector<int> visited;

int find(int x){
    if(parent[x] != x){
        return find(parent[x]);
    }
    return x;
}

void union_set(int a, int b){
    a = find(a);
    b = find(b);

    if(a < b){
        parent[b] = a;  
    }else{
        parent[a] = b;
    }
}

int main(){
    cin >> n;
    cin >> m;

    for(int i=0; i<n+1;i++){
        parent.push_back(i);
    }

    for(int i=0; i < n; i++){
        for(int j=0; j<n; j++){
            cin >> a;
            if(a == 1){
                union_set(i+1,j+1);
            }
        }
    }

    int tmp;
    for(int i=0; i<m;i++){
        cin >> tmp;
        visited.push_back(tmp);
    }


    bool check = true;
    int current = -1;
    for(int i = 0; i< m; i++){
        if(i==0){
            if(parent[visited[i]] != visited[i]){
                parent[visited[i]] = find(visited[i]);
            }
            current = parent[visited[i]];
        }
        else{
        if(parent[visited[i]] != visited[i]){
            parent[visited[i]] = find(visited[i]);
        }
        if(parent[visited[i]] != current){
            check = false;
            break;
        }

        }
    }

    if(check == true){
        cout << "YES";
    }else{
        cout << "NO";
    }

}


